phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -420,8 +421,11 @@ public ConnectionQueryServicesImpl(QueryServices services, ConnectionInfo connec
             list.add(queue);
         }
         connectionQueues = ImmutableList.copyOf(list);
+
         // A little bit of a smell to leak `this` here, but should not be a problem
-        this.tableStatsCache = new GuidePostsCache(this, config);
+        this.tableStatsCache = queryServicesHelper.getGuidePostsCache(props.get(GUIDE_POSTS_CACHE_FACTORY_CLASS,
+                QueryServicesOptions.DEFAULT_GUIDE_POSTS_CACHE_FACTORY_CLASS), this, config);
```

Comment:
The reference is getting escaped here in the constructor. Should we discuss if there is a better way to initialize this? May be `CQSI#init()`?

Reply:
Just curious, may be we don't even need it?

Reply:
So I spent most of my time looking at this and deciding if I can/should address this now.  After discussion I want to make a new JIRA and address in a separate PR.  This change for this escaping likely needs to go all the way to the PhoenixDriver level to properly handle this creation as well as disentangle the queryServices and the stats cache.  Are you okay with this @karanmehta93 ?

Reply:
As discussed offline I made https://issues.apache.org/jira/browse/PHOENIX-5244 to track this.